### DISCUSSION TITLE
Querying Vulnerabilities for Particular Version Ranges

### DISCUSSION THREAD

**Discussion by giftcup** [google/osv.dev#1852] [General]
How would you query for vulnerabilities that are related to a range of versions for of an npm package? 

For example, if you have a package listed as follows; "@apollo/client": "^3.0.0" in the package.json file, how would you query for all vulnerabilities that are related to the package's versions using the OSV api?

**Comment by michaelkedar** [comment]
Hi,
Currently, we don't support querying version ranges directly in the API. Generally, we expect people to be looking for vulnerabilities in installed package, so you could query the installed version in the package-lock.json file.

If you need to query the version range, you can omit the `version` field in the query API request to get every vulnerability that affects any version of the package, then manually parse the [`affected.ranges`](https://ossf.github.io/osv-schema/#affectedranges-field) fields for matches of the version range.
So the query for your example would be simply:
```json
{
  "package": {
    "name": "@apollo/client",
    "ecosystem": "npm"
  }
}
```

May I ask what your use-case is that you need to do this?

**Comment by andrewpollock** [comment]
@giftcup circling back on the request for more detail on your particular use case

**Comment by giftcup** [comment]
I was trying to build a tool that would detect vulnerable packages in a package.json file without creating the package-lock.json. I thought I could query the package with the version ranges like those listed on a package.json file.

Sorry for the late reply.

**Comment by andrewpollock** [comment]
Disclaimer: I'm fairly ignorant about NPM, and I'm basing this on the most convenient `package.json` file I have available: https://github.com/google/osv.dev/blob/2c31244e38fb6e9bc53653a4406ad14765c62a9f/gcp/appengine/frontend3/package.json#L1-L39

Can you not simply call the OSV API with the package and versions present here?

e.g.

```
$ curl -d \
  '{"version": "1.3.0",
    "package": {"name": "@github/clipboard-copy-element", "ecosystem": "npm"}}' \
  "https://api.osv.dev/v1/query"
```

@giftcup You've mentioned "ranges" (plural) here, but I'm not seeing how you'd have more than a single version to be querying for? Do you have a `package.json` file example you can provide?
